src/ol_openedx_canvas_integration/api.py

```diff
@@ -0,0 +1,220 @@
+"""Utility functions for canvas integration"""
+import logging
+from collections import defaultdict
+
+from opaque_keys.edx.locator import CourseLocator
+
+from lms.djangoapps.courseware.courses import get_course_by_id
+from lms.djangoapps.grades.course_grade_factory import CourseGradeFactory
+from common.djangoapps.student.models import CourseEnrollmentAllowed, CourseEnrollment
+from ol_openedx_canvas_integration.client import CanvasClient, create_assignment_payload, update_grade_payload_kv
+from django.contrib.auth.models import User
+from lms.djangoapps.courseware.access import has_access
+from lms.djangoapps.grades.context import grading_context_for_course
+
+log = logging.getLogger(__name__)
+
+
+def first_or_none(iterable):
+    """Returns the first item in the given iterable, or None if the iterable is empty"""
+    return next((x for x in iterable), None)
+
+
+def course_graded_items(course):
+    grading_context = grading_context_for_course(course)
+    for graded_item_type, graded_items in grading_context['all_graded_subsections_by_type'].items():
+        for graded_item_index, graded_item in enumerate(graded_items, start=1):
+            yield graded_item_type, graded_item, graded_item_index
+
+
+def get_enrolled_non_staff_users(course):
+    """
+    Returns an iterable of non-staff enrolled users for a given course
+    """
+    return [
+        user for user in CourseEnrollment.objects.users_enrolled_in(course.id)
+        if not has_access(user, 'staff', course)
+    ]
+
+
+def enroll_emails_in_course(emails, course_key):
+    """
+    Attempts to enroll all provided emails in a course. Emails without a corresponding
+    user have a CourseEnrollmentAllowed object created for the course.
+    """
+    results = {}
+    for email in emails:
+        user = User.objects.filter(email=email).first()
+        result = ''
+        if not user:
+            _, created = CourseEnrollmentAllowed.objects.get_or_create(
+                email=email,
+                course_id=course_key
+            )
+            if created:
+                result = 'User does not exist - created course enrollment permission'
+            else:
+                result = 'User does not exist - enrollment is already allowed'
+        elif not CourseEnrollment.is_enrolled(user, course_key):
+            try:
+                CourseEnrollment.enroll(user, course_key)
+                result = 'Enrolled user in the course'
+            except Exception as ex:  # pylint: disable=broad-except
+                result = 'Failed to enroll - {}'.format(ex)
+        else:
+            result = 'User already enrolled'
+        results[email] = result
+    return results
+
+
+def get_subsection_user_grades(course):
```

Comment:
@arslanashraf7 I think we can remove a lot of code around getting user's grate by assignment and use builtin- [method](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/lms/djangoapps/grades/models_api.py#L78) of PersistentGrades

Reply:
As discussed keeping the scope of this in mind, I've created a ticket to keep track of this optimization https://github.com/mitodl/open-edx-plugins/issues/10.